gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -798,99 +804,26 @@ case class ColumnarOverrideRules(session: SparkSession)
-    if (queryFallbackThreshold < 0) {
-      return false
-    }
-    countFallbacks(plan) >= queryFallbackThreshold
-  }
-
-  private def countFallbacks(plan: SparkPlan): Int = {
-    var fallbacks = 0
-    def countFallback(plan: SparkPlan): Unit = {
-      plan match {
-        // Another stage.
-        case _: QueryStageExec =>
-          return
-        case ColumnarToRowExec(p: GlutenPlan) =>
-          logDebug(s"c2r: $p")
-          fallbacks = fallbacks + 1
-        // Possible fallback for leaf node.
-        case leafPlan: LeafExecNode if !leafPlan.isInstanceOf[GlutenPlan] =>
-          fallbacks = fallbacks + 1
-        case _ =>
-      }
-      plan.children.foreach(p => countFallback(p))
-    }
-    countFallback(plan)
-    fallbacks
-  }
-
-  /**
-   * Ported from ApplyColumnarRulesAndInsertTransitions of Spark. Inserts an transition to columnar
-   * formatted data.
-   */
-  private def insertRowToColumnar(plan: SparkPlan): SparkPlan = {
-    if (!plan.supportsColumnar) {
-      // The tree feels kind of backwards
-      // Columnar Processing will start here, so transition from row to columnar
-      RowToColumnarExec(insertTransitions(plan, outputsColumnar = false))
-    } else if (!plan.isInstanceOf[RowToColumnarTransition]) {
-      plan.withNewChildren(plan.children.map(insertRowToColumnar))
-    } else {
-      plan
-    }
-  }
-
-  /**
-   * Ported from ApplyColumnarRulesAndInsertTransitions of Spark. Inserts RowToColumnarExecs and
-   * ColumnarToRowExecs where needed.
-   */
-  def insertTransitions(plan: SparkPlan, outputsColumnar: Boolean): SparkPlan = {
-    if (outputsColumnar) {
-      insertRowToColumnar(plan)
-    } else if (plan.supportsColumnar) {
-      // `outputsColumnar` is false but the plan outputs columnar format, so add a
-      // to-row transition here.
-      ColumnarToRowExec(insertRowToColumnar(plan))
-    } else if (!plan.isInstanceOf[ColumnarToRowTransition]) {
-      plan.withNewChildren(plan.children.map(insertTransitions(_, outputsColumnar = false)))
-    } else {
-      plan
-    }
-  }
-
-  // Just for test use.
-  def enableAdaptiveContext(): Unit = {
-    isAdaptiveContext = true
-  }
-
   override def postColumnarTransitions: Rule[SparkPlan] = plan =>
     PhysicalPlanSelector.maybe(session, plan) {
-      if (fallbackPolicy == "query" && !isAdaptiveContext && fallbackWholeQuery(plan)) {
-        logWarning("Fall back to run the query due to unsupported operator!")
-        insertTransitions(originalPlan, outputsColumnar = false)
-      } else if (fallbackPolicy == "stage" && isAdaptiveContext && fallbackWholeStage(plan)) {
-        logWarning("Fall back the plan due to meeting the whole stage fallback threshold!")
-        insertTransitions(originalPlan, outputsColumnar = false)
-      } else {
-        transformPlan(postOverrides(), plan, "post")
+      val planWithFallbackPolicy = transformPlan(fallbackPolicy(), plan, "fallback")
```

Comment:
Just out of my curiosity, would it be possible to apply `fallbackPolicy()` in preColumnarTransitions?

Reply:
There are two things need to do if we want to move `fallbackPolicy()` to preColumnarTransitions.
- We count the c2r inside `fallbackPolicy()`, and c2r only happens in postColumnarTransitions
- We want to skip `postOverrides` if we decide to fallback

So seems it's simpler to do it at postColumnarTransitions

Reply:
> We count the c2r inside fallbackPolicy(), and c2r only happens in postColumnarTransitions

Would it be feasible to count on the conjective non-transformable tags?

I think C2R is unlikely reliable to indicate fallback: it should be based on the assumption that Vanilla Spark is always row-based, while Gluten Spark is always not. However the assumption can be breaked since it's somewhat implicit. For example sometime Vanilla Spark has columanr processing either, some day Gluten may have the ability to fallback row processing to native, etc.

I saw you added "InsertTransitions" which does the similar job comparing to vanilla Spark's. Moving the logic to preColumnarTransitions can leave that job to vanilla Spark.

You can consider the comment non-blocking since it requires a lot of changes. Since the PR already makes things clearer than before.

Reply:
conjective non-transformable looks good to me, and we can even support partial fallback in one stage. We'd better to consider the cost of each operator. e.g., if one stage has a non-transformable project and several transformable joins, shall we fallack it.